dallinger/experiment_server/dashboard.py

```diff
@@ -652,6 +661,229 @@ def dashboard_lifecycle():
+    - Cleaned and formatted for display
+    - Sent as an SSE message
+
+    Returns
+    -------
+    Response
+        A Flask response object with mimetype 'text/event-stream' containing
+        the event stream of log updates. Each event contains:
+        - message : str
+            The formatted log message with ANSI codes converted to HTML
+        - original_line : str
+            The raw log line before processing
+        - log_line_number : int, optional
+            The line number in the log file
+
+    Notes
+    -----
+
+    The client should connect to this endpoint using an EventSource object
+    to receive the updates.
+    """
+
+    def generate():
+        for line in Pygtail(LOG_FILE):
+            try:
+                line_dict = clean_line_info(json.loads(line))
+                line_dict["original_line"] = line
+                yield f"data:{json.dumps(line_dict)}\n\n"
+            except json.decoder.JSONDecodeError:
+                yield f"data:{json.dumps({'message': line})}\n\n"
+
+    return Response(generate(), mimetype="text/event-stream")
+
+
+@dashboard.route("/logs/range", methods=["GET"])
+@login_required
+def logs_range():
+    """
+    Return a range of log lines from the log file.
+    """
+    params = request.args
+    start, end = params.get("start", None), params.get("end", None)
+    if start is None or end is None:
+        return json_error_response("Both 'start' and 'end' parameters are required.")
+    start, end = int(start), int(end)
+    if start < 1:
+        return json_error_response("'start' must be > 0.")
+
+    if start > end:
+        return json_error_response("'start' must be <= 'end'.")
+
+    lines, early_stop, last_line = log_read_lines(start, end + 1)
+    if not early_stop and start > last_line:
+        return json_error_response(
+            f"'start' must be less than the current line number {last_line}."
+        )
+
+    return lines
+
+
+def json_error_response(message, status_code=400):
+    return jsonify({"msg": message}), status_code
+
+
+@dashboard.route("/logs/find_lines", methods=["GET"])
+@login_required
+def logs_find_lines():
+    """
+    Find lines in the log file that contain a substring (GET parameter 'query').
+    """
+    params = request.args
+    query = params.get("query", None)
+    if query is None:
+        return json_error_response("No query provided.")
+    return Response(log_search_substring(query), mimetype="text/event-stream")
+
+
+@dashboard.route("/logs/find_line_number", methods=["POST"])
+@login_required
+def logs_find_line_number():
+    """
+    Find the line number in the log file that contains a substring (POST parameter 'query').
+    """
+    params = request.json
+    query = params.get("query", None)
+    if query is None:
+        return json_error_response("No query provided.")
+
+    line_number = find_log_line_number(query)
+    if line_number is not None:
+        return jsonify({"line_number": line_number})
```

Comment:
Are you sure you need `jsonify` here? I thought this is done automatically by Flask.